### ISSUE TITLE
frontend: hard errors alerts as ratio/percentage instead of absolute-value

### ISSUE THREAD

**Issue by bobheadxi** [sourcegraph/sourcegraph#12158]
Writeup by @slimsag in https://github.com/sourcegraph/sourcegraph/issues/12011 re: `frontend: hard_error_search_responses`

---

It's not clear to me if this one is a fluke or not, because [sourcegraph.com has some quite bad search perf issues we haven't been following up on](https://github.com/sourcegraph/sourcegraph/issues/9359) - but a more direct query shows me the percentage of requests resulting in a hard error here is low so I presume we do not care and would like to relax the alert to account for a percentage instead of an absolute number of failing requests.

[this query](https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-7d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum%20by%20(source,request_name)(increase(src_graphql_search_response%7Bstatus%3D~%5C%22error%5C%22%7D%5B1d%5D))%22,%22legendFormat%22:%22hard%20error%22,%22datasource%22:%22Prometheus%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D&right=%5B%22now-7d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum%20by%20(source,request_name)(increase(src_graphql_search_response%7Bstatus!~%5C%22error%5C%22%7D%5B1d%5D))%22,%22legendFormat%22:%22hard%20error%22,%22datasource%22:%22Prometheus%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D)

shows that the cause is about 300 browser code intel requests per day (find-references, jump-to-def, etc.) end in a hard error, where the user gets an error back instead of a result. If I had to guess, these are something like a timeout edge case we aren't recording properly (e.g. the user navigates away from the page while it is in flight), or correlated with deployments of services.

While that is worrying, the same query above shows that only accounts for only around 0.1% of requests soI think the TODO here is two-fold:

1. File an issue to track turning this into a ratio/percentage instead of absolute-value alert.
2. Silence the alert for now on Sourcegraph.com

**Comment by bobheadxi** [comment]
Unclear if this should be @sourcegraph/distribution or @sourcegraph/cloud . Tagging it as cloud for now, feel free to comment and change!

**Comment by keegancsmith** [comment]
I agree, this should be a ratio. FYI this alert dates back to when we first added search and an absolute value was more effective due to periods of no searches.

I think you can be quite confident for these sorts of changes to just do them, and justify in PR description :)

**Comment by keegancsmith** [comment]
Another thought: if you are an instance with low search volume. A search failing may trigger the % error in the last 5 min. So if we do add an alert based on %, it should also have a minimum total number of events in the 5min before firing.

**Comment by keegancsmith** [comment]
@slimsag can I encode this in our monitoring system? This is something I could do when directly using prometheus alerts, but not sure when looking at our monitoring DSL if I can? Looks like the DSL would have to be extended to support this sort of query.

Additionally previously we would create recording rules to help create more readable alerts queries. It would be great to get that back. I may have to do some hacking :)

**Comment by keegancsmith** [comment]
To capture some context from @bobheadxi and I's hacking earlier. Ideally we could use prometheus's `for` keyword on alerts to help increase signal on this alert once it is a ratio.

There is the general problem of how we do alerts for small instances vs large. Small instances can have periods of low/no traffic leading to ratio based alerts falsely firing (we experienced this a lot when I first added prometheus to sourcegraph). For smaller instances sustained periods (30m?) of bad error ratio is likely a good enough reason to raise the alert. For larger installations we can alert more quickly. Not exactly sure how best to resolve this.
